Tests/SourceKitLSPTests/SemanticTokensTests.swift

```diff
@@ -0,0 +1,689 @@
+  }
+
+  private func performSemanticTokensRequest(text: String, range: Range<Position>? = nil) -> [Token] {
+    let response: DocumentSemanticTokensResponse!
+
+    if let range = range {
+      response = try! sk.sendSync(DocumentSemanticTokensRangeRequest(textDocument: textDocument, range: range))
+    } else {
+      response = try! sk.sendSync(DocumentSemanticTokensRequest(textDocument: textDocument))
+    }
+
+    return [Token](lspEncodedTokens: response.data)
+  }
+
+  private func openAndPerformSemanticTokensRequest(text: String, range: Range<Position>? = nil) -> [Token] {
+    openDocument(text: text)
+    return performSemanticTokensRequest(text: text, range: range)
+  }
+
+  func testIntArrayCoding() {
+    let tokens = [
+      Token(
+        start: Position(line: 2, utf16index: 3),
+        utf16length: 5,
+        kind: .string
+      ),
+      Token(
+        start: Position(line: 4, utf16index: 2),
+        utf16length: 1,
+        kind: .interface,
+        modifiers: [.deprecated, .definition]
+      ),
+    ]
+
+    let encoded = tokens.lspEncoded
+    XCTAssertEqual(encoded, [
+      2, // line delta
+      3, // char delta
+      5, // length
+      Token.Kind.string.rawValue, // kind
+      0, // modifiers
+
+      2, // line delta
+      2, // char delta
+      1, // length
+      Token.Kind.interface.rawValue, // kind
+      Token.Modifiers.deprecated.rawValue | Token.Modifiers.definition.rawValue, // modifiers
+    ])
+
+    let decoded = [Token](lspEncodedTokens: encoded)
+    XCTAssertEqual(decoded, tokens)
+  }
+
+  func testRangeSplitting() {
+    let text = """
+    struct X {
+      let x: Int
+      let y: String
+
+
+    }
+    """
+    openDocument(text: text)
+
+    guard let snapshot = connection.server?.workspace?.documentManager.latestSnapshot(uri) else {
+      fatalError("Could not fetch document snapshot for \(#function)")
+    }
+
+    let empty = Position(line: 0, utf16index: 1)..<Position(line: 0, utf16index: 1)
+    XCTAssertEqual(empty._splitToSingleLineRanges(in: snapshot, omittingEmptyRange: true), [])
+    XCTAssertEqual(empty._splitToSingleLineRanges(in: snapshot, omittingEmptyRange: false), [empty])
+
+    for omittingEmptyRange in [false, true] {
+      let multiLine = Position(line: 1, utf16index: 6)..<Position(line: 2, utf16index: 7)
+      XCTAssertEqual(multiLine._splitToSingleLineRanges(in: snapshot, omittingEmptyRange: omittingEmptyRange), [
+        Position(line: 1, utf16index: 6)..<Position(line: 1, utf16index: 12),
+        Position(line: 2, utf16index: 0)..<Position(line: 2, utf16index: 7),
+      ])
+
+      let emptyLines = Position(line: 2, utf16index: 14)..<Position(line: 5, utf16index: 1)
+      XCTAssertEqual(emptyLines._splitToSingleLineRanges(in: snapshot, omittingEmptyRange: omittingEmptyRange), [
+        Position(line: 2, utf16index: 14)..<Position(line: 2, utf16index: 15),
+        Position(line: 3, utf16index: 0)..<Position(line: 3, utf16index: 0),
```

Comment:
Just to confirm that this is intentional - we would generate a SemanticToken for this, right?

Reply:
IIRC the end position here is exclusive, so having them be the same doesn't quite make sense, right?

Reply:
Good catch, thanks!